### PR TITLE
Add design doc for training to explain accelerate inclusion for FSDP

### PR DESCRIPTION
**Pull request by Maxusmusti**
A write-up explaining the decision to include Hugging Face `accelerate` as a dependency for our implementation of FSDP in `instructlab-training`, as well as a breakdown of the considerations made, detailing both the benefits and risks to consider when introducing new Hugging Face dependencies.

### CODE REVIEW

File: docs/training/training-accelerate-dep.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+For example, if we wanted to change how model saving worked, or a user wanted to simply understand how model saving worked, they would no longer be able to do so through our library directly, but instead would need to read HF documentation, and contribute back to HF repos with the hope that the customization is considered generally useful and approved. More niche or experimental customizations may not be approved at all, and would have to be patched in directly or overwritten in a rather hacky manner within our library. This as described is not the most dire of circumstances, as one could argue that if it is not ready to be included upstream, it may not be ready to be included in our library. Where things get more serious, however, is with the introduction of bugs. At this very moment, there is a bug in `v0.34.2` that means FSDP model saving does not behave as expected. This fix is resolved in `v1.0.0rc1`, but there is no official release yet that can be pulled in with this fix. So for RHEL AI 1.2, we are currently maintaining a patch directly in our library to fix the bug while working with `v0.34.2`, and it will sit there temporarily until Accelerate has a new official release.
```
Comment (leseb) [inline docs/training/training-accelerate-dep.md:109]: Isn't this inherent to all usage of a library?

Comment (Maxusmusti) [reply docs/training/training-accelerate-dep.md:109]: Not necessarily, this is the inclusion of a library that sits directly on top of and obfuscates the internal implementation that was previously exposed directly within our library. So it is inherent to any library being pulled in as an abstraction layer for previously directly-implemented code, but a new library/dependency can be pulled in for other reasons like additional functionality, separate code paths, etc. That being said, you are correct that this is a very widely-applicable and usually assumed claim, but it had to be re-iterated for the people that were curious as to why we didn't just pull in ALL HF libraries and try to abstract away everything 😅 (you'd be surprised how often we've received this question lol)

**Review by jaideepr97 — APPROVED** [review]
just a couple nits
Thanks for writing this @Maxusmusti !

File: docs/training/training-accelerate-dep.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+Ultimately, these risks and overhead management costs make sense for Accelerate, because the package provides significant benefit that far outweigh our concerns. When considering additional HF packages, however, there is currently **no obvious benefit** to their inclusion, which only further annunciates the risks associated. It must be made clear that the risks of Accelerate are non-negligible, and that the conveniences and positive impact provided are what push it over the line of inclusion. It is by no means a "no brainer" to further include any additional HF packages at this time, unless they afford us a similar benefit.
```
[the remark this answers is not in the document]
Comment (Maxusmusti) [reply docs/training/training-accelerate-dep.md:123]: Yes this is what was meant, thanks

File: docs/training/training-accelerate-dep.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+## The immediate benefit
+
+The inclusion of Accelerate in this manner drastically simplifies the process of supporting multiple sharding frameworks. Rather than having diverging code paths for model setup, optimizer setup, sharding configuration, distributed initialization, state saving, and model checkpointing, all of these training steps can be supported with the same code. The common abstraction allows us to maintain both DeepSpeed and FSDP, and prepares us for inclusion of additional sharding frameworks.
+
+## Impact on overhead
+
+### Performance
+
+There has been no noticeable performance impact observed during development, but we will defer to the Performance and Scale team for final measurements.
```
[the remark this answers is not in the document]
Comment (Maxusmusti) [reply docs/training/training-accelerate-dep.md]: yeah makes sense, removed mention

**Comment by mairin** [comment]
Just a quick q @Maxusmusti - this pulls in accelerate as a new dependency - does it have any dependencies it pulls in further from that?

**Comment by Maxusmusti** [comment]
@mairin Nothing else, just accelerate. All other dependencies (and in fact accelerate as well) for FSDP were already being pulled in, we just wanted to make sure our reasoning behind accelerate was clear since it was now being explicitly required by the training library and becoming a more major inclusion in code.

**Pull request merged**